File: AppSets-iOS/AppSets/com/app/appsets/server/model/UserInfo.swift
//
//  UserInfo.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/4.
//

import Foundation
import RealmSwift

final class UserInfo : Object, ImSessionHolder, Bio, Codable {

    @Persisted(primaryKey: true)
    var uid: String
    @Persisted
    var name: String? = nil
    @Persisted
    var avatarUrl: String? = nil
    @Persisted
    var agreeToTheAgreement: Int? = nil
    @Persisted
    var age: Int? = nil
    @Persisted
    var sex: String? = nil
    @Persisted
    var email: String? = nil
    @Persisted
    var phone: String? = nil
    @Persisted
    var address: String? = nil
    @Persisted
    var introduction: String? = nil
    @Persisted
    var company: String? = nil
    @Persisted
    var profession: String? = nil
    @Persisted
    var website: String? = nil
    @Persisted
    var roles: String? = nil
    
    var bioUrl: String? {
        get{
            return avatarUrl
        }
        set{
            
        }
    }
    
    var id: String {
        get {
            return uid
        }
    }
    
    var session: Session? = nil
    
    
    enum CodingKeys: String, CodingKey {
        case uid
        case name
        case avatarUrl
        case agreeToTheAgreement
        case age
        case sex
        case email
        case phone
        case address
        case introduction
        case company
        case profession
        case website
    }
    
    override init() {
        
    }
    
    init(uid:String) {
        super.init()
        self.uid = uid
    }
    
    
    init(
        uid:String,
        name:String? = nil,
        age:Int? = nil,
        sex:String? = nil
    ) {
        super.init()
        self.uid = uid
        self.name = name
        self.age = age
        self.sex = sex
    }
    
    init(from decoder: any Decoder) throws {
        super.init()
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.uid = try container.decode(String.self, forKey: .uid)
        self.name = try container.decodeIfPresent(String.self, forKey: .name)
        self.avatarUrl = try container.decodeIfPresent(String.self, forKey: .avatarUrl)
        self.agreeToTheAgreement = try container.decodeIfPresent(Int.self, forKey: .agreeToTheAgreement)
        let ageString = try container.decodeIfPresent(String.self, forKey: .age)
        if ageString != nil {
            if let ageInt = Int(ageString!) {
                self.age = ageInt
            }
        }
        
        self.sex = try container.decodeIfPresent(String.self, forKey: .sex)
        self.email = try container.decodeIfPresent(String.self, forKey: .email)
        self.phone = try container.decodeIfPresent(String.self, forKey: .phone)
        self.address = try container.decodeIfPresent(String.self, forKey: .address)
        self.introduction = try container.decodeIfPresent(String.self, forKey: .introduction)
        self.company = try container.decodeIfPresent(String.self, forKey: .company)
        self.profession = try container.decodeIfPresent(String.self, forKey: .profession)
        self.website = try container.decodeIfPresent(String.self, forKey: .website)
    }
    
    func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.uid, forKey: .uid)
        try container.encodeIfPresent(self.name, forKey: .name)
        try container.encodeIfPresent(self.avatarUrl, forKey: .avatarUrl)
        try container.encodeIfPresent(self.agreeToTheAgreement, forKey: .agreeToTheAgreement)
        try container.encodeIfPresent("\(String(describing: self.age))", forKey: .age)
        try container.encodeIfPresent(self.sex, forKey: .sex)
        try container.encodeIfPresent(self.email, forKey: .email)
        try container.encodeIfPresent(self.phone, forKey: .phone)
        try container.encodeIfPresent(self.address, forKey: .address)
        try container.encodeIfPresent(self.introduction, forKey: .introduction)
        try container.encodeIfPresent(self.company, forKey: .company)
        try container.encodeIfPresent(self.profession, forKey: .profession)
        try container.encodeIfPresent(self.website, forKey: .website)
    }
    
    private static let user0 = UserInfo(uid: "0", name: "蒋开心", age: 0, sex: "female")
    
    static func defaultUser() -> UserInfo {
        return user0
    }
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/server/model/UserInfo.kt
package xcj.app.appsets.server.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import xcj.app.appsets.im.Bio

@Entity(tableName = "UserInfo")
data class UserInfo(
    /**
     * @hidden
     */
    @PrimaryKey
    var uid: String,
    var name: String? = null,
    var age: Int? = null,
    var sex: String? = null,
    var email: String? = null,
    var phone: String? = null,
    var address: String? = null,
    @ColumnInfo(name = "avatar_url")
    var avatarUrl: String? = null,
    var introduction: String? = null,
    var company: String? = null,
    var profession: String? = null,
    var website: String? = null,
    var roles: String? = null,
    @Ignore
    var agreeToTheAgreement: Int? = null,
) : Bio {

    override val bioId: String
        get() = uid

    override val bioName: String?
        get() = name

    @Ignore
    override var bioUrl: Any? = null

    constructor() : this(
        "",
        null,
        null,
        Int.MIN_VALUE.toString(),
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null
    )

    fun isDefault(): Boolean {
        return this.uid == user0.bioId
    }

    companion object {
        private val user0 = UserInfo(
            "U0",
            "蒋开心",
            18,
            "female",
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
        )

        fun default(): UserInfo {
            return user0
        }

        fun basic(uid: String, name: String?, avatarUrl: String?): UserInfo {
            return UserInfo(uid = uid, name = name, avatarUrl = avatarUrl)
        }

        fun isContentSame(a: UserInfo, other: UserInfo): Boolean {
            return a.agreeToTheAgreement == other.agreeToTheAgreement &&
                    a.uid == other.uid &&
                    a.name == other.name &&
                    a.age == other.age &&
                    a.sex == other.sex &&
                    a.email == other.email &&
                    a.phone == other.phone &&
                    a.address == other.address &&
                    a.avatarUrl == other.avatarUrl &&
                    a.introduction == other.introduction &&
                    a.company == other.company &&
                    a.profession == other.profession &&
                    a.website == other.website
        }
    }
}